kafka/protocol: use default generated operator==()

tchaikov opened the pull request:
## Cover letter

simpler this way. also, this helps the compiler to find the comparison
operators before they are used.

this change helps to address the FTBFS with clang++-15 + libstd++-12:

/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_algobase.h:1161:22: error: invalid operands to binary expression ('const kafka::api_versions_response_key' and 'const kafka::api_versions_response_key')
            if (!(*__first1 == *__first2))
                  ~~~~~~~~~ ^  ~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_algobase.h:1210:38: note: in instantiation of function template specialization 'std::__equal<false>::equal<const kafka::api_versions_response_key *, const kafka::api_versions_response_key *>' requested here
      return std::__equal<__simple>::equal(__first1, __last1, __first2);
                                     ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_algobase.h:1218:19: note: in instantiation of function template specialization 'std::__equal_aux1<const kafka::api_versions_response_key *, const kafka::api_versions_response_key *>' requested here
      return std::__equal_aux1(std::__niter_base(__first1),
                  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_algobase.h:1555:19: note: in instantiation of function template specialization 'std::__equal_aux<__gnu_cxx::__normal_iterator<const kafka::api_versions_response_key *, std::vector<kafka::api_versions_response_key>>, __gnu_cxx::__normal_iterator<const kafka::api_versions_response_key *, std::vector<kafka::api_versions_response_key>>>' requested here
      return std::__equal_aux(__first1, __last1, __first2);
                  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_vector.h:2037:16: note: in instantiation of function template specialization 'std::equal<__gnu_cxx::__normal_iterator<const kafka::api_versions_response_key *, std::vector<kafka::api_versions_response_key>>, __gnu_cxx::__normal_iterator<const kafka::api_versions_response_key *, std::vector<kafka::api_versions_response_key>>>' requested here
              && std::equal(__x.begin(), __x.end(), __y.begin())); }
                      ^
/home/kefu/dev/redpanda/src/v/kafka/protocol/api_versions.h:67:35: note: in instantiation of function template specialization 'std::operator==<kafka::api_versions_response_key, std::allocator<kafka::api_versions_response_key>>' requested here
               && a.data.api_keys == b.data.api_keys
                                  ^
/home/kefu/dev/redpanda/src/v/kafka/protocol/join_group.h:137:13: note: candidate function not viable: no known conversion from 'const kafka::api_versions_response_key' to 'const std::vector<join_group_request_protocol>' for 1st argument
inline bool operator==(
            ^

Signed-off-by: Kefu Chai <tchaikov@gmail.com>


## Release notes

* none

senior7515 commented:
this is great @tchaikov 

cc: @graphcareful @mmaslankaprv

tchaikov commented on src/v/kafka/protocol/schemata/generator.py:
```diff
@@ -845,6 +850,15 @@ def value_type(self):
         return snake_case(self._field["name"])
 
+    @property
+    def has_default_three_way_comparision(self):
```
okay. just thought that the comparison operators were synthesized from the shuttle operator. but yeah, `is_default_comparable` is way shorter. =)

tchaikov commented:
implemented the changeset with the "allowlist" solution and repushed.

tchaikov commented on src/v/kafka/protocol/schemata/generator.py:
```diff
@@ -845,6 +850,15 @@ def value_type(self):
+    @property
+    def has_default_three_way_comparision(self):
+        # assume that other types can be compared
+        if self.is_array:
```
yeah. i should have checked the doc more closely.

senior7515 commented:
> so how about just adding a list for the types we want to have `operator=()`, namely `api_versions_response_key`, `supported_feature_key` and `finalized_feature_key`?

@tchaikov what about having an "allow-list" of types for which we will gen the operator= - or maybe a table ref, so folks can opt-into it.

graphcareful commented on src/v/kafka/protocol/api_versions.h:
```diff
@@ -60,34 +60,8 @@ struct api_versions_response final {
-               && a.data.api_keys == b.data.api_keys
-               && a.data.supported_features == b.data.supported_features
-               && a.data.finalized_features == b.data.finalized_features;
+        return a.data == b.data;
```
this method can be defaulted as well

tchaikov replied:
done.

graphcareful commented on src/v/kafka/protocol/api_versions.h:
```diff
@@ -57,37 +57,7 @@ struct api_versions_response final {
-               && a.data.supported_features == b.data.supported_features
-               && a.data.finalized_features == b.data.finalized_features;
-    }
+    bool operator==(const kafka::api_versions_response&) = default;
```
Can be a hidden friend too

tchaikov replied:
sure, fixed.

tchaikov commented on src/v/kafka/protocol/schemata/generator.py:
```diff
@@ -893,6 +900,9 @@ def name(self):
     {{ struct.context_field[0] }} {{ struct.context_field[1] -}};
 {%- endif %}
+{%- if struct.needs_equality_operator %}
+    bool operator==(const {{ struct.name }}&) const = default;
```
@BenPope thought 
```c++
bool operator==(const foo&) const = default;
```
and 
```c++
friend bool operator==(const foo&, const foo&) = default;
```

are equivalent, am i right? and i'd prefer the former, as it's shorter and hence needs less typing =)

graphcareful replied:
There are explicit benefits of hidden friends, off the top of my head avoiding accidental implicit conversions. I found a good article about it here: https://www.justsoftwaresolutions.co.uk/cplusplus/hidden-friends.html

BenPope replied:
This is unfortunately not true: ([example on godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxCBmAKykAA6oCoRODB7evnrJqY4CQSHhLFEx8baY9vkMQgRMxASZPn5cFVXptfUEhWGR0bEJCnUNTdmtQ109xaUDAJS2qF7EyOwc5gDMwcjeWADUJutuTkPEmKwH2CYaAIIbWzuY%2B4dUXgwO6WIXV7c3J14OuwQeHQWAYAH0qMQ8K59gB2Kw3XZI3bBAi7ABuBwR12Ru0h0IY6F2EVQnl2qES0SYRGIBwAIvSIGgGENAcDQRCoa5zAA2Ui7ZmsoEgxicgnoXmzJ503ZYKhMLy0AhY74mWEM9bY75/AFsErROHY3GojEqxHIklkilUmn0xmCtF60qSgUCVn02WYeWK5Wa1Xqs0/a4mlhMYIQKVqo3I4Uc/EwhAaKNJgN%2B81I2Oi%2BOEwFcKN51PRpFDdAgFCLNEHNxV3YQRPSj0ILizQO4ktltBeSuHGsQdsgU5UOsaSPrDUy/uDuvNltpoO4p0GlgpiwpjVF3aL4ibgsWAvr75tgil8tdp7Vw615cNsc72cb/ud7sXty1ydeiDL0fj3bvocsGdAzVBkbg4eZaE4OJeD8DgtFIVBOGrSxrF/RZlkeDYeFIAhNDA%2BYAGsQDiSQADo4h5SRJDiABOaj1g0dYuHWGi%2BQgjhJF4FgJA0DRSBguCEI4XgFBAXicNgsDSDgWAYEQcsWESOhonISg0AUpSYm2QxgDBAhiFefC%2BDoAhohEiAIlw0gImCeoAE9OCwtS2EEAB5BhaHsiTSCwUMjHELz8FOd50UwESvMwVRMGQLtVjg1FKgc3haDwCJiDsjwsEsvS8C4iT5ioAxgAUAA1aEAHcXMpGCsP4QQRDEdgpBkQRFBUdQvN0VoDCMFBrGsfQUpEyB5gpaowoAWhLelTGQywzA0XZJuPZKInpFaZqsSxhMqKLqhcQlRhaUhAmCXp9RiVpcjSARDpyFJroYKY%2Bgutpdo6YZGk8Zo9DsN6BE6BonvOn6Ptu8YPqBmZmwWJYVj0PTMFWHhwMg6DLME1QAA4eXGiiBW64Baz0gypQgJDNpsXZcEIEh9jMRj%2BQ8dT6G3TDZl4cStFmeYEDOLAYgjUhCLiDQSNhTGzC4DR6eorhYVhKWEjYjjSC49Z1hI9Wte1nW%2BX43hBOE0TsNw%2BZpLktTFJZlSmVQZn%2BiK5hEgUBBUAIIylVMygLK86zmGITzHLt5yCDcjzLJ87T/LgwK3pCsK4IiqKYsS8hBAS7gkpStKA4y2KOahXKub4QqSvKyrGFT2rhFEcQmur1q1Es3QzH0bTetmmwVqGwXRvSCaprHDbrHm7b2mcCBXDB47CUh/pLvu6pp6u6o55e373n%2B0GvrGV7N5qCHTumefbG3rIjomQGj%2BeiR5gUNC4daBGkcktioL49HOCxnG8eAZBkCJvpBg%2BFSbk36lTfANI6YM12EzK2BoNhmHZibPKPM%2Bb9EFsrTiIB1aax1vg9WetP5CVsMbTmklzZ1hABWRIXYbaWw0qEVgqxv640kLsP%2BACIDE2AcgzAkCSDAj0NXeqddpANyUE3DqsRSBlTSokRKKMODv31vBTgLkuy0LRKgKguxWG/3/oAkmtY4EaWgUgjmpslEqzVhrAhBCP5eUNqQsSVihYgCoiRSQzFiI8R5NRHk2NJCtzYusNGTjOCWNQfoTgZhwkCUiSgrm8wQrEFSM4SQQA%3D))

graphcareful replied:
@BenPope , try this example instead: https://godbolt.org/z/csM4e4o7G to my knowledge this is an example of ADL vs normal lookup

tchaikov replied:
@graphcareful thank you Rob! reading through it.

graphcareful commented:
> > I believe this change is a good idea, but the implementation should change. Off the top of my head an idea would be to hardcode types in the generator source that we know are not comparable (like `partition_produce_data`), and just don't define default equality operators for those types (and types that have these types nested) statically ahead of time.
> 
> @graphcareful thanks for your review! yeah, the `optional<>` and `vector<>` rules are wrong. as the standard library defines the comparison operators for them with some restrictions, so under most circumstances, these types are comparable.
> 
> but, i am not sure if we want to add `operator==()` for _most_ (if not all of them) of the generated kafka types. for couple reasons,
> 
> * if we don't use it, we should not pay for it. as quite a few equity operators would never be used. shall we ask the compiler for them? this'd incur build-time (slower compilation) and runtime (larger executable size) overheads.
> * technically, if one of the members is not default-comparable, the compiler would not be able to define the shuttle operator for the container struct. and this policy is enforced recursively. as you pointed out above. so without building a set of types which are not `is_default_comparable`, and updating the set when walking through the types, the `operator bool(...) = default` statement would be a wrong direction sometimes.
> 
> so how about just adding a list for the types we want to have `operator=()`, namely `api_versions_response_key`, `supported_feature_key` and `finalized_feature_key`?
> 
> i know, this is not as sustainable as the solution which uses a "bad" types list. but it is at least correct and introduces minimum build-time and runtime overhead.
> 
> what do you think?

I thought that too when first developing, I asked @BenPope his ideas, he told me these were invalid concerns, linker does a good job at this. To the point about compile times, we'd only be defining the method for a few hundrew structs, not thousands or anything super large. We could measure the impact to find out if these are valid concerns.

I also do like the idea of an equality operator for each struct, because I plan to add robust testing, a serde test for each request at all supported versions, and i'd be leveraging the equality operator in most if not all structs in this case.

tchaikov commented on src/v/kafka/protocol/schemata/generator.py:
```diff
@@ -634,6 +633,13 @@ def structs(self):
 
+    @property
+    def needs_equality_operator(self):
+        return self.name in ('api_versions_response_data',
```
done.

tchaikov commented:
> 
> I thought that too when first developing, I asked @BenPope his ideas, he told me these were invalid concerns, linker does a good job at keeping binary size small. To the point about compile times, we'd only be defining the method for a few hundred structs, not thousands or anything super large. We could measure the impact to find out if these are valid concerns.
> 
> I also do like the idea of an equality operator for each struct, because I plan to add robust testing, a serde test for each request at all supported versions, and i'd be leveraging the equality operator in most if not all structs in this case.

agreed. the equality operator would be handy in this use case. will try to implement the "ineligible" list solution this weekend.

graphcareful commented:
> > I thought that too when first developing, I asked @BenPope his ideas, he told me these were invalid concerns, linker does a good job at keeping binary size small. To the point about compile times, we'd only be defining the method for a few hundred structs, not thousands or anything super large. We could measure the impact to find out if these are valid concerns.
> > I also do like the idea of an equality operator for each struct, because I plan to add robust testing, a serde test for each request at all supported versions, and i'd be leveraging the equality operator in most if not all structs in this case.
> 
> agreed. the equality operator would be handy in this use case. will try to implement the "ineligible" list solution this weekend.

It will be awesome to review !

tchaikov commented:
@graphcareful @BenPope hi Rob and Ben, could you please help re-review this changeset? i think all comments have been addressed.

tchaikov commented:
changes since the previous revision

- add an "ineligible" list for the "native" types which do not have default-generated `operator==()`
- use hidden `friend operator==()` instead.
- add a helper decorator for caching the `is_default_comparable` property of the generated field types otherwise we need to do a depth first traversal for each of the struct types. 
- bring `is_default_comparable()` back to field and struct types.
- generate `bool operator==` for every struct as long as it `is_default_comparable` .

graphcareful commented on src/v/kafka/protocol/schemata/generator.py:
```diff
@@ -436,6 +436,27 @@ def make_context_field(path):
+
+# cache a property of a type by its name if it would be expensive to
+# recalculate this property
+def cache_by_name(user_func):
```
This seems like an over-optimization. This file takes no time to run.

graphcareful commented on src/v/kafka/protocol/schemata/generator.py:
```diff
@@ -436,6 +436,27 @@ def make_context_field(path):
 
+# a list of struct types which are ineligible to have default-generated
+# `operator==()`, because one or more of its member variables cannot do
+# three-way comparsion.
```
instead of `three-way` you can say `are not comparable`

tchaikov replied:
sure.

graphcareful reviewed (changes requested):
I think this is g2g with these minor edits considered

tchaikov commented:
change log

- dropped the `cache` decorator
- s/three way comparison/comparable/

BenPope commented:
@tchaikov looks ok to me, but it's failing the lint checks.

tchaikov commented:
> @tchaikov looks ok to me, but it's failing the lint checks.

fixed and repushed.

graphcareful reviewed (approved):
LGTM

graphcareful commented:
Thanks, @tchaikov great contribution !

dotnwat reviewed (commented):
lgtm
